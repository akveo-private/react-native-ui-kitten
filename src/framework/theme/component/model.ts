// TODO(@type): declare theme types

export type ThemeShape = any;
export type ThemedStyleShape = any;
export type StyleSheetShape = any;
