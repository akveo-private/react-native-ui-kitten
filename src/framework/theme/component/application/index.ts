export {
  ApplicationProvider,
  Props as ApplicationProviderProps,
} from './applicationProvider.component';
