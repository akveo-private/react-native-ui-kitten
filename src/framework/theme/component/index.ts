export { ThemeProvider } from './themeProvider.component';
export { withTheme } from './themeConsumer.component';
export { withThemedStyles } from './themeConsumerStyled.component';
export { ThemeType } from './type';
