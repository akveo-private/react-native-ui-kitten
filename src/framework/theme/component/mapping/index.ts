export * from './type';
export {
  MappingProvider as ThemeMappingProvider,
  Props as ThemeMappingProviderProps,
} from './mappingProvider.component';
