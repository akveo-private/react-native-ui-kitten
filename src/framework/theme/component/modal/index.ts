export * from './modal.component';
export * from './modalPanel.component';
