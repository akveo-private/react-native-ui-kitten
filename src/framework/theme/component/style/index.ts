export {
  StyleProvider,
  Props as StyleProviderProps,
} from './styleProvider.component';

export {
  styled,
  Props as StyledComponentProps,
} from './styleConsumer.component';
