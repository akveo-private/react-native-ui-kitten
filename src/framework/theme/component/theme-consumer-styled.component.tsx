import React from 'react';
import { StyleSheet } from 'react-native';
import {
  ThemeType,
  ThemedStyleType,
  StyleSheetType,
} from './type';
import {
  Consumer,
  forwardProps,
} from '../service';

export interface Props<P> extends React.ClassAttributes<P> {
  theme: ThemeType;
  themedStyle: ThemedStyleType;
}

export function withThemedStyles<P extends Props<P>>(
  Component: React.ComponentType<P>,
  createStyles: (theme: ThemeType) => StyleSheetType,
) {
  type TExcept = Exclude<keyof P, keyof Props<P>>;
  type ForwardedProps = Pick<P, TExcept>;

  class Shadow extends React.Component<ForwardedProps> {
    wrappedComponentRef = undefined;
    getWrappedInstance = undefined;

    setWrappedComponentRef = (ref) => {
      this.wrappedComponentRef = ref;
    };

    renderWrappedComponent = (theme: ThemeType) => {
      const styles = StyleSheet.create(createStyles(theme));
      return (
        <Component
          ref={this.setWrappedComponentRef}
          theme={theme}
          themedStyle={styles}
          {...this.props}
        />
      );
    };

    render() {
      return (
        <Consumer>
          {this.renderWrappedComponent}
        </Consumer>
      );
    }
  }

  const Result = Shadow;
  Result.prototype.getWrappedInstance = function getWrappedInstance() {
    const hasWrappedInstance = this.wrappedComponentRef && this.wrappedComponentRef.getWrappedInstance;
    return hasWrappedInstance ? this.wrappedComponentRef.getWrappedInstance() : this.wrappedComponentRef;
  };

  return forwardProps(Component, Result);
}
