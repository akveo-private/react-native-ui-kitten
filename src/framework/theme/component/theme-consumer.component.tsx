import React from 'react';
import {
  Consumer,
  forwardProps,
} from '../service';

export interface ThemedComponentProps<T> extends React.ClassAttributes<T> {
  theme: Object;
}

export function withTheme<T extends ThemedComponentProps<T>>(Component: React.ComponentType<T>) {
  type TExcept = Exclude<keyof T, keyof ThemedComponentProps<T>>;
  type ForwardedProps = Pick<T, TExcept>;

  class ConsumingComponent extends React.Component<ForwardedProps, {}> {
    rootComponentRef = undefined;

    getConsumingComponent = undefined;

    constructor(props) {
      super(props);
      this.setRootComponentRef = this.setRootComponentRef.bind(this);
    }

    setRootComponentRef(ref) {
      this.rootComponentRef = ref;
    }

    renderRootComponent(component: React.ComponentType<T>, theme: Object) {
      const RootComponent = component;
      return (
        <RootComponent
          ref={this.setRootComponentRef}
          theme={theme}
          {...this.props}
        />
      );
    }

    render() {
      return (
        <Consumer>
          {theme => this.renderRootComponent(Component, theme)}
        </Consumer>
      );
    }
  }

  const ResultComponent = ConsumingComponent;
  ResultComponent.prototype.getConsumingComponent = function getWrappedInstance() {
    const hasWrappedInstance = this.rootComponentRef.getConsumingComponent;
    return hasWrappedInstance ? this.rootComponentRef.getConsumingComponent() : this.rootComponentRef;
  };

  return forwardProps(ResultComponent, Component);
}
