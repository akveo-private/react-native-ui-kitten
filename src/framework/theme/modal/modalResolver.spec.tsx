import React from 'react';
import {
  View,
  Text,
  Button,
  TouchableOpacity,
} from 'react-native';
import {
  fireEvent,
  render,
  RenderAPI,
  shallow,
} from 'react-native-testing-library';
import { ModalResolver } from './modalResolver.component';

jest.useFakeTimers();

describe('@modal resolver component checks', () => {

  const MODAL_TEST_IDENTIFIER = (substring: string): string => {
    return `modal-test-identifier-${substring}`;
  };

  interface TestScreenProps {
    buttonTestId: string;
  }

  interface TestScreenState {
    modalVisible: boolean;
  }

  class TestScreen extends React.Component<TestScreenProps, TestScreenState> {

    public state: TestScreenState = {
      modalVisible: false,
    };

    private setModalVisible(modalVisible: boolean): void {
      this.setState({ modalVisible });
    }

    public render(): React.ReactNode {
      return (
        <View>
          <Button
            testID={this.props.buttonTestId}
            title='Open Modal'
            onPress={() => this.setModalVisible(true)}/>
          <ModalResolver
            visible={this.state.modalVisible}
            isBackDropAllowed={true}
            identifier={MODAL_TEST_IDENTIFIER('1')}
            onCloseModal={() => 1}>
            <View>
              <Text>
                Test2
              </Text>
            </View>
          </ModalResolver>,
        </View>
      );
    }
  }

  it('* modal resolver component renders properly', () => {
    const modal1: RenderAPI = render(
      <ModalResolver
        visible={true}
        isBackDropAllowed={false}
        identifier={MODAL_TEST_IDENTIFIER('1')}
        onCloseModal={() => 1}>
        <View>
          <Text>
            Test1
          </Text>
        </View>
      </ModalResolver>,
    );

    const modal2: RenderAPI = render(
      <ModalResolver
        visible={false}
        isBackDropAllowed={false}
        identifier={MODAL_TEST_IDENTIFIER('1')}
        onCloseModal={() => 1}>
        <View>
          <Text>
            Test2
          </Text>
        </View>
      </ModalResolver>,
    );

    const { output: firstOutput } = shallow(modal1.getByType(ModalResolver));
    const { output: secondOutput } = shallow(modal2.getByType(ModalResolver));

    expect(firstOutput).toMatchSnapshot();
    expect(secondOutput).toMatchSnapshot();
  });

  it('* modal resolver component props checks', () => {
    const modalPassingProps = {
      visible: true,
      isBackDropAllowed: false,
    };
    const modal = <ModalResolver {...modalPassingProps}/>;

    expect(modal.props.visible).toBe(modalPassingProps.visible);
    expect(modal.props.isBackDropAllowed).toBe(modalPassingProps.isBackDropAllowed);
  });

  it('* modal resolver closes on passed prop', () => {
    const onCloseModal = jest.fn();

    const component: RenderAPI = render(
      <ModalResolver
        visible={true}
        isBackDropAllowed={true}
        identifier={MODAL_TEST_IDENTIFIER('1')}
        onCloseModal={onCloseModal}>
        <View>
          <Text>Test1</Text>
          <Button
            title='Close Modal'
            onPress={onCloseModal}
          />
        </View>
      </ModalResolver>,
    );


    const { output: openedOutput } = shallow(component.getByType(ModalResolver));
    expect(openedOutput).toMatchSnapshot();

    fireEvent.press(component.getByType(Button));
    expect(onCloseModal).toHaveBeenCalled();

    const { output: closedOutput } = shallow(component.getByType(ModalResolver));
    expect(closedOutput).toMatchSnapshot();
  });

  it('* modal resolver component close on backDrop checks', () => {
    const onCloseModal = jest.fn();

    const component: RenderAPI = render(
      <ModalResolver
        visible={true}
        isBackDropAllowed={true}
        identifier={MODAL_TEST_IDENTIFIER('1')}
        onCloseModal={onCloseModal}>
        <View>
          <Text>
            Test1
          </Text>
        </View>
      </ModalResolver>,
    );

    const { output: openedOutput } = shallow(component.getByType(ModalResolver));
    expect(openedOutput).toMatchSnapshot();

    fireEvent.press(component.getByType(TouchableOpacity));
    expect(onCloseModal).toHaveBeenCalled();

    const { output: closedOutput } = shallow(component.getByType(ModalResolver));
    expect(closedOutput).toMatchSnapshot();
  });

  it('* component styled with mappings', () => {
    const component: RenderAPI = render(
      <ModalResolver visible={true}>
        <Text>Test</Text>
      </ModalResolver>,
    );

    const { output } = shallow(component.getByType(ModalResolver));

    expect(output).toMatchSnapshot();
  });

});
