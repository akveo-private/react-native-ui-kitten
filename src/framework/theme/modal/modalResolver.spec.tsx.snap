// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@modal resolver component checks * component styled with mappings 1`] = `
<View
  style={
    Object {
      "height": 0,
      "position": "absolute",
      "width": 0,
    }
  }
>
  <View
    pointerEvents="none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    pointerEvents="box-none"
    style={
      Object {
        "alignSelf": "flex-start",
      }
    }
    visible={true}
  >
    <Text
      onCloseModal={[Function]}
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`@modal resolver component checks * modal resolver closes on passed prop 1`] = `
<TouchableOpacity
  activeOpacity={1}
  onPress={[Function]}
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    identifier="modal-test-identifier-1"
    onCloseModal={[MockFunction]}
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    pointerEvents="box-none"
    style={
      Object {
        "alignSelf": "flex-start",
      }
    }
    visible={true}
  >
    <View
      onCloseModal={[Function]}
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <Text>
        Test1
      </Text>
      <Button
        onPress={[MockFunction]}
        title="Close Modal"
      />
    </View>
  </View>
</TouchableOpacity>
`;

exports[`@modal resolver component checks * modal resolver closes on passed prop 2`] = `
<TouchableOpacity
  activeOpacity={1}
  onPress={[Function]}
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    identifier="modal-test-identifier-1"
    onCloseModal={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    pointerEvents="box-none"
    style={
      Object {
        "alignSelf": "flex-start",
      }
    }
    visible={true}
  >
    <View
      onCloseModal={[Function]}
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <Text>
        Test1
      </Text>
      <Button
        onPress={
          [MockFunction] {
            "calls": Array [
              Array [
                undefined,
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        title="Close Modal"
      />
    </View>
  </View>
</TouchableOpacity>
`;

exports[`@modal resolver component checks * modal resolver component close on backDrop checks 1`] = `
<TouchableOpacity
  activeOpacity={1}
  onPress={[Function]}
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    identifier="modal-test-identifier-1"
    onCloseModal={[MockFunction]}
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    pointerEvents="box-none"
    style={
      Object {
        "alignSelf": "flex-start",
      }
    }
    visible={true}
  >
    <View
      onCloseModal={[Function]}
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <Text>
        Test1
      </Text>
    </View>
  </View>
</TouchableOpacity>
`;

exports[`@modal resolver component checks * modal resolver component close on backDrop checks 2`] = `
<TouchableOpacity
  activeOpacity={1}
  onPress={[Function]}
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    identifier="modal-test-identifier-1"
    onCloseModal={
      [MockFunction] {
        "calls": Array [
          Array [
            "modal-test-identifier-1",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    pointerEvents="box-none"
    style={
      Object {
        "alignSelf": "flex-start",
      }
    }
    visible={true}
  >
    <View
      onCloseModal={[Function]}
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <Text>
        Test1
      </Text>
    </View>
  </View>
</TouchableOpacity>
`;

exports[`@modal resolver component checks * modal resolver component renders properly 1`] = `
<View
  style={
    Object {
      "height": 0,
      "position": "absolute",
      "width": 0,
    }
  }
>
  <View
    pointerEvents="none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    identifier="modal-test-identifier-1"
    onCloseModal={[Function]}
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    pointerEvents="box-none"
    style={
      Object {
        "alignSelf": "flex-start",
      }
    }
    visible={true}
  >
    <View
      onCloseModal={[Function]}
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <Text>
        Test1
      </Text>
    </View>
  </View>
</View>
`;

exports[`@modal resolver component checks * modal resolver component renders properly 2`] = `null`;
